server/data/data-library/src/main/java/org/apache/james/domainlist/lib/AbstractDomainList.java

```diff
@@ -62,12 +71,40 @@ protected Logger getLogger() {
 
     @Override
     public void configure(HierarchicalConfiguration config) throws ConfigurationException {
-        defaultDomain = config.getString("defaultDomain", "localhost");
+        setDefaultDomain(config);
 
         setAutoDetect(config.getBoolean("autodetect", true));
         setAutoDetectIP(config.getBoolean("autodetectIP", true));
     }
 
+    @VisibleForTesting void setDefaultDomain(HierarchicalConfiguration config) throws ConfigurationException {
+        
+        try {
+            defaultDomain = config.getString("defaultDomain", LOCALHOST);
+
+            String hostName = InetAddress.getLocalHost().getHostName();
+            if (mayChangeDefaultDomain()) {
+                setDefaultDomain(hostName);
+            }
+        } catch (UnknownHostException e) {
+            LOGGER.warn("Unable to retrieve hostname.", e);
+        } catch (DomainListException e) {
+            LOGGER.error("An error occured while creating the default domain", e);
+        }
+    }
+
+    private boolean mayChangeDefaultDomain() {
+        return !Strings.isNullOrEmpty(defaultDomain)
```

Comment:
The `AbstractConfiguration.getString(String, String)` may return empty string, so i added this test.